cmd/server/main.go

```diff
@@ -39,80 +38,64 @@ type Arguments struct {
+	// Check if USB device is available
+	// TODO(jeremy): add search filter for product and model info to confirm rplidar is present instead of assuming
+	usbDevices := usb.Search(
+		usb.SearchFilter{},
+		func(vendorID, productID int) bool {
+			return true
+		})
+
+	if len(usbDevices) != 0 {
+		golog.Global.Debugf("detected %d lidar devices", len(usbDevices))
+		for _, comp := range usbDevices {
 			golog.Global.Debug(comp)
 		}
-	}
-	if argsParsed.DevicePath != "" {
-		lidarComponents = []config.Component{{Type: config.ComponentTypeLidar, Model: rplidar.ModelName, Host: argsParsed.DevicePath}}
+	} else {
+		return errors.New("no usb devices found")
 	}
 
-	if len(lidarComponents) == 0 {
-		return errors.New("no lidar devices found")
+	// Create rplidar component
+	lidarDevice := config.Component{
+		Name:       "rplidar",
+		Type:       config.ComponentTypeCamera,
+		Model:      rplidar.ModelName,
+		Attributes: config.AttributeMap{"device_path": "/dev/ttyUSB0"},
 	}
 
-	lidarComponent := lidarComponents[0]
-	if lidarComponent.Model != rplidar.ModelName {
-		return errors.New("device is not rplidar")
+	// Add device path if specified
+	if argsParsed.DevicePath != "" {
+		lidarDevice.Attributes = config.AttributeMap{"device_path": argsParsed.DevicePath}
 	}
 
-	return runServer(ctx, int(argsParsed.Port), lidarComponent, logger)
+	return runServer(ctx, int(argsParsed.Port), lidarDevice, logger)
 }
 
 func runServer(ctx context.Context, port int, lidarComponent config.Component, logger golog.Logger) (err error) {
-	r, err := robotimpl.New(ctx, &config.Config{Components: []config.Component{lidarComponent}}, logger)
-	if err != nil {
-		return err
-	}
-	lidarDevice := r.LidarByName(r.LidarNames()[0])
-
-	info, err := lidarDevice.Info(ctx)
-	if err != nil {
-		return err
-	}
-	golog.Global.Infow("rplidar", "info", info)
-	defer func() {
-		err = multierr.Combine(err, lidarDevice.Stop(context.Background()))
-	}()
 
-	listener, err := net.Listen("tcp", fmt.Sprintf("localhost:%d", port))
+	metadataSvc, err := service.New()
 	if err != nil {
 		return err
 	}
+	ctx = service.ContextWithService(ctx, metadataSvc)
 
-	rpcServer, err := rpcserver.New(logger)
+	cfg := &config.Config{Components: []config.Component{lidarComponent}}
+	myRobot, err := robotimpl.New(ctx, cfg, logger, client.WithDialOptions(rpc.WithInsecure()))
 	if err != nil {
 		return err
 	}
-	defer func() {
-		err = multierr.Combine(err, rpcServer.Stop())
-	}()
-
-	if err := rpcServer.RegisterServiceServer(
-		ctx,
-		&pb.RobotService_ServiceDesc,
-		grpcserver.New(r),
-		pb.RegisterRobotServiceHandlerFromEndpoint,
-	); err != nil {
-		return err
-	}
 
-	go func() {
-		<-ctx.Done()
-		if err := rpcServer.Stop(); err != nil {
-			panic(err)
-		}
-	}()
-	utils.ContextMainReadyFunc(ctx)()
-	return rpcServer.Serve(listener)
+	options := web.NewOptions()
+	options.Network = config.NetworkConfig{BindAddress: fmt.Sprintf(":%d", port)}
+	return webserver.RunWeb(ctx, myRobot, options, logger)
```

Comment:
Note: Kat experienced issues closing out of the server needing to resort to CTRL-\ to do it. Since the server is basically be held by the webserver.RunWeb at this point, my thought that this indicates a potential issue with it and not this cmd/server/main.go repo. But I could be mistaken.

Reply:
The issue was that there was a workgroup added to the code that:
a) was not needed
b) was blocking the `Close` function from finishing.